lib/Transforms/AffineDataCopyGeneration.cpp

```diff
@@ -625,35 +639,48 @@ LogicalResult AffineDataCopyGeneration::runOnBlock(Block *block) {
                                                  : exceedsCapacity(forOp);
       if (recurseInner) {
         // We'll recurse and do the copies at an inner level for 'forInst'.
-        runOnBlock(/*begin=*/curBegin, /*end=*/it);
         // Recurse onto the body of this loop.
         runOnBlock(forOp.getBody());
-        // The next block range starts right after the 'affine.for' operation.
-        curBegin = std::next(it);
       } else {
         // We have enough capacity, i.e., copies will be computed for the
         // portion of the block until 'it', and for 'it', which is 'forOp'. Note
         // that for the latter, the copies are placed just before this loop (for
         // incoming copies) and right after (for outgoing ones).
-        runOnBlock(/*begin=*/curBegin, /*end=*/it);
 
         // Inner loop copies have their own scope - we don't thus update
         // consumed capacity. The footprint check above guarantees this inner
         // loop's footprint fits.
         runOnBlock(/*begin=*/it, /*end=*/std::next(it));
-        curBegin = std::next(it);
       }
-    } else if (!isa<AffineLoadOp>(&*it) && !isa<AffineStoreOp>(&*it)) {
+      // Get to the next load or store op after 'forOp'.
+      curBegin = std::find_if(std::next(it), block->end(), [&](Operation &op) {
+        return (isa<AffineLoadOp>(op) || isa<AffineStoreOp>(op) ||
+                isa<AffineForOp>(op)) &&
+               copyNests.count(&op) == 0;
+      });
+      it = curBegin;
+    } else if (copyNests.count(&*it) > 0) {
+      // This op is a copy nest and the range ends here.
       runOnBlock(/*begin=*/curBegin, /*end=*/it);
-      curBegin = std::next(it);
+      // Get to the start of the next range.
+      curBegin = std::find_if(std::next(it), block->end(), [&](Operation &op) {
+        return (isa<AffineLoadOp>(op) || isa<AffineStoreOp>(op) ||
+                isa<AffineForOp>(op)) &&
+               copyNests.count(&op) == 0;
+      });
+      it = curBegin;
+    } else {
+      // We simply include this op in the current range and continue for more.
+      ++it;
```

Comment:
These changes look fine, but this function (and loop here) is getting quite long. It would be nice to have a follow up PR that broken this pass up into functions and refactored things a bit...

Reply:
Yes, I'll do that next. I actually have some changes ready that refactor this pass code into a separate data copy generation utility moved into lib/Transforms/Utils/Utils.cpp - so that it's reusable elsewhere. I'm waiting for this to get in.

Reply:
I've now turned this into a while loop (didn't make a lot of sense as a 'for'). This loop is a one-pass on a block of op's and it's processing (contiguous) ranges of op's from among it - while skipping over portions. So it's really not a structured one. It's now shorter with one of the cases eliminated - I'll see if this could be made more readable in the next change on this.